### PR TITLE
User Content progress details

### PR DESCRIPTION
**Pull request by pc0202**

### CODE REVIEW

File: course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala [hunk 1]
```diff
@@ -27,6 +27,20 @@ class ContentConsumptionActor @Inject() extends BaseEnrolmentActor {
...
+        add("userid")
+        add("batchid")
+        add("courseid")
+        add("completedcount")
+        add("completionpercentage")
+        add("lastcompletedtime")
+        add("progress")
+        add("status")
+        add("viewcount")
+    }}
```
Comment (SanthoshVasabhaktula) [inline course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala:41]: @pritha-tarento - Any reason why default fields have been added? and if they are needed, can the default fields be taken up from a configuration file?

File: course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala [hunk 2]
```diff
@@ -27,6 +27,20 @@ class ContentConsumptionActor @Inject() extends BaseEnrolmentActor {
...
+    val progressDetails = "progressDetails" //this is related to the response/request key
+    val progressdetails = "progressdetails" //this maps to DB column
```
Comment (SanthoshVasabhaktula) [inline course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala:43]: @pritha-tarento - do not hardcode JSON parsing only for progress details. Please create a list property named as "jsonFields" and convert to & from json for all the fields in this list

File: course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala [hunk 3]
```diff
@@ -177,22 +191,24 @@ class ContentConsumptionActor @Inject() extends BaseEnrolmentActor {
...
+        val response = cassandraOperation.getRecords(requestContext, consumptionDBInfo.getKeySpace, consumptionDBInfo.getTableName, filters, fields)
         response.getResult.getOrDefault(JsonKey.RESPONSE, new java.util.ArrayList[java.util.Map[String, AnyRef]]).asInstanceOf[java.util.List[java.util.Map[String, AnyRef]]]
     }
 
     def processContentConsumption(inputContent: java.util.Map[String, AnyRef], existingContent: java.util.Map[String, AnyRef], userId: String) = {
         val inputStatus = inputContent.getOrDefault(JsonKey.STATUS, 0.asInstanceOf[AnyRef]).asInstanceOf[Number].intValue()
         val updatedContent = new java.util.HashMap[String, AnyRef]()
         updatedContent.putAll(inputContent)
+        if(inputContent.get(progressDetails) != null)
+            updatedContent.put(progressDetails, mapper.writeValueAsString(inputContent.get(progressDetails)))
```
Comment (SanthoshVasabhaktula) [inline course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala:211]: @pritha-tarento - Refer to my earlier comment and generalize this code.

File: course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala [hunk 4]
```diff
@@ -293,12 +309,25 @@ class ContentConsumptionActor @Inject() extends BaseEnrolmentActor {
         val courseId = request.get(JsonKey.COURSE_ID).asInstanceOf[String]
         val contentIds = request.getRequest.getOrDefault(JsonKey.CONTENT_IDS, new java.util.ArrayList[String]()).asInstanceOf[java.util.List[String]]
         val fields = request.getRequest.getOrDefault(JsonKey.FIELDS, new java.util.ArrayList[String](){{ add(JsonKey.PROGRESS) }}).asInstanceOf[java.util.List[String]]
-        val contentsConsumed = getContentsConsumption(userId, courseId, contentIds, batchId, request.getRequestContext)
+        //default fields added ..
+        if (fields!=null && fields.contains(progressDetails) && !defaultFields.contains(progressdetails)){
+            defaultFields.add(progressdetails)
+        } else if(fields!=null && !fields.contains(progressdetails)){
+            defaultFields.remove(progressdetails)
+        }
```
Comment (SanthoshVasabhaktula) [inline course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala:317]: @pritha-tarento - this code is not required

File: course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala [hunk 5]
```diff
@@ -293,12 +309,25 @@ class ContentConsumptionActor @Inject() extends BaseEnrolmentActor {
...
         val response = new Response
         if(CollectionUtils.isNotEmpty(contentsConsumed)) {
             val filteredContents = contentsConsumed.map(m => {
                 ProjectUtil.removeUnwantedFields(m, JsonKey.DATE_TIME, JsonKey.USER_ID, JsonKey.ADDED_BY, JsonKey.LAST_UPDATED_TIME)
                 m.put(JsonKey.COLLECTION_ID, m.getOrDefault(JsonKey.COURSE_ID, ""))
+                //added progress starts....
+                if(fields.contains(progressDetails)){
+                    m.put(progressDetails, mapper.readTree(m.get(progressdetails).asInstanceOf[String]))
+                    m.remove(progressdetails)
+                }
```
Comment (SanthoshVasabhaktula) [inline course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala:329]: @pritha-tarento - Same as the comment earlier. Generalize this code

File: course-mw/sunbird-util/sunbird-cassandra-utils/src/main/java/org/sunbird/cassandraimpl/CassandraDACImpl.java [hunk 6]
```diff
@@ -32,7 +32,7 @@ public Response getRecords(
     try {
       Select select;
       if (CollectionUtils.isNotEmpty(fields)) {
-        select = QueryBuilder.select((String[]) fields.toArray()).from(keySpace, table);
+        select = QueryBuilder.select(fields.toArray()).from(keySpace, table);
```
Comment (SanthoshVasabhaktula) [inline course-mw/sunbird-util/sunbird-cassandra-utils/src/main/java/org/sunbird/cassandraimpl/CassandraDACImpl.java:35]: @pritha-tarento - Any reason for this change? Can this reverted back if the change is not required?

**Review by SanthoshVasabhaktula — CHANGES_REQUESTED** [review]

File: course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala (hunk 5 above)
Comment (pritha-tarento) [reply course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala]: Resolved

File: course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala (hunk 4 above)
Comment (pritha-tarento) [reply course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala]: Resolved

File: course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala (hunk 3 above)
Comment (pritha-tarento) [reply course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala]: Resolved

File: course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala (hunk 2 above)
Comment (pritha-tarento) [reply course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ContentConsumptionActor.scala]: Apply changes to generalized

File: course-mw/sunbird-util/sunbird-cassandra-utils/src/main/java/org/sunbird/cassandraimpl/CassandraDACImpl.java (hunk 6 above)
Comment (pritha-tarento) [reply course-mw/sunbird-util/sunbird-cassandra-utils/src/main/java/org/sunbird/cassandraimpl/CassandraDACImpl.java:35]: This change is required. Scenario is when the required fields passed as parameter this throws exception to not able to pull the records from underlying DB.

**Review by SanthoshVasabhaktula — CHANGES_REQUESTED** [review]
@pritha-tarento - please update all data types to scala collection types.

**Pull request closed without merging**